src/src/dynpower.cpp

```diff
@@ -0,0 +1,202 @@
+    // state == connected is not used: unplugging an RX will be connected and will boost power to max before disconnect
+    if (armed && (powerHeadroom > 0))
+    {
+      uint32_t linkstatsInterval = ExpressLRS_currTlmDenom * ExpressLRS_currAirRate_Modparams->interval / (1000U / 2U);
+      linkstatsInterval = std::max(linkstatsInterval, (uint32_t)512U);
+      if ((now - dynpower_last_linkstats_millis) > (linkstatsInterval + 2U))
+      {
+        DBGLN("+power (tlm)");
+        POWERMGNT::incPower();
+      }
+    }
+    return;
+  }
+
+  // If no new telemetry, no new LQ/SNR/RSSI to use for adjustment
+  if (!newTlmAvail)
+    return;
+  dynpower_last_linkstats_millis = now;
+
+  // =============  LQ-based power boost up ==============
+  // Quick boost up of power when detected any emergency LQ drops.
+  // It should be useful for bando or sudden lost of LoS cases.
+  uint32_t lq_current = CRSF::LinkStatistics.uplink_Link_quality;
+  uint32_t lq_avg = dynpower_mavg_lq;
+  int32_t lq_diff = lq_avg - lq_current;
+  dynpower_mavg_lq.add(lq_current);
+  // if LQ drops quickly (DYNPOWER_LQ_BOOST_THRESH_DIFF) or critically low below DYNPOWER_LQ_BOOST_THRESH_MIN, immediately boost to the configured max power.
+  if (lq_diff >= DYNPOWER_LQ_BOOST_THRESH_DIFF || lq_current <= DYNPOWER_LQ_BOOST_THRESH_MIN)
+  {
+      DynamicPower_SetToConfigPower();
+      return;
+  }
+
+  PowerLevels_e startPowerLevel = POWERMGNT::currPower();
+  if (ExpressLRS_currAirRate_RFperfParams->DynpowerSnrThreshUp == DYNPOWER_SNR_THRESH_NONE)
+  {
+    // =============  RSSI-based power increment ==============
+    // a simple threshold compared against N sample average of
+    // rssi vs the sensitivity limit +/- some thresholds
+    dynpower_mean_rssi.add(rssi);
+
+    if (dynpower_mean_rssi.getCount() >= DYNPOWER_RSSI_CNT)
+    {
+      int32_t expected_RXsensitivity = ExpressLRS_currAirRate_RFperfParams->RXsensitivity;
+      int8_t rssi_inc_threshold = expected_RXsensitivity + DYNPOWER_RSSI_THRESH_UP;
+      int8_t rssi_dec_threshold = expected_RXsensitivity + DYNPOWER_RSSI_THRESH_DN;
+      int8_t avg_rssi = dynpower_mean_rssi.mean(); // resets it too
+      if ((avg_rssi < rssi_inc_threshold) && (powerHeadroom > 0))
+      {
+        DBGLN("+power (rssi)");
+        POWERMGNT::incPower();
+      }
+      else if (avg_rssi > rssi_dec_threshold && lq_avg >= DYNPOWER_LQ_THRESH_DN)
+      {
+        DBGVLN("-power (rssi)");
+        POWERMGNT::decPower();
+      }
+    }
+  } // ^^ if RSSI-based
+  else
+  {
+    // =============  SNR-based power increment ==============
+    // Decrease the power if SNR above threshold and LQ is good
+    // Increase the power for each (X) SNR below the threshold
+    //int8_t snrScaled = CRSF::LinkStatistics.uplink_SNR;
+    if (snrScaled >= ExpressLRS_currAirRate_RFperfParams->DynpowerSnrThreshDn && lq_avg >= DYNPOWER_LQ_THRESH_DN)
+    {
+      DBGVLN("-power (snr)");
```

Comment:
Also, here, DBGVLN for down, an DBGLN for up. So I assume you intend this to be the case.

Reply:
I do remember now you mention it. When I was typing the comments I felt a little déjà vu moment in my head.
Good comments in the code, I won't ask why it's like this again 😉